File: project_euler/problem_155/solution.py
"""
Problem 155: Count distinct capacitance values using up to 18 unit capacitors.

S(1) = {1}
S(n) = union over k=1..n//2 of { a+b, a*b/(a+b) : a in S(k), b in S(n-k) }
C(n) = |S(1) union ... union S(n)|

Uses tuples (numerator, denominator) for speed instead of Fraction.
"""
from math import gcd

def make_frac(p, q):
    """Return reduced fraction as (p, q) with q > 0."""
    if q < 0:
        p, q = -p, -q
    g = gcd(abs(p), q)
    return (p // g, q // g)

def solve():
    MAXN = 18

    S = [set() for _ in range(MAXN + 1)]
    S[1].add((1, 1))

    all_values = {(1, 1)}

    for n in range(2, MAXN + 1):
        for k in range(1, n // 2 + 1):
            m = n - k
            for (ap, aq) in S[k]:
                for (bp, bq) in S[m]:
                    # Parallel: a + b
                    pn = ap * bq + bp * aq
                    pd = aq * bq
                    S[n].add(make_frac(pn, pd))

                    # Series: ab/(a+b)
                    sn = ap * bp
                    sd = ap * bq + bp * aq
                    S[n].add(make_frac(sn, sd))

        all_values |= S[n]
        print(f"n={n}: |S(n)|={len(S[n])}, C(n)={len(all_values)}")

    print(len(all_values))

solve()

File: project_euler/problem_155/solution.cpp
#include <bits/stdc++.h>
using namespace std;

/*
 * Problem 155: Count distinct capacitance values using up to 18 unit capacitors.
 *
 * S(1) = {1}
 * S(n) = union over k=1..n/2 of { a+b, ab/(a+b) : a in S(k), b in S(n-k) }
 * C(n) = |S(1) union ... union S(n)|
 */

typedef pair<int, int> Frac; // (numerator, denominator), always reduced, den > 0

Frac make_frac(long long p, long long q) {
    if (q < 0) { p = -p; q = -q; }
    long long g = __gcd(abs(p), q);
    return {(int)(p / g), (int)(q / g)};
}

int main() {
    const int MAXN = 18;

    vector<set<Frac>> S(MAXN + 1);
    S[1].insert({1, 1});

    set<Frac> all_values;
    all_values.insert({1, 1});

    for (int n = 2; n <= MAXN; n++) {
        for (int k = 1; k * 2 <= n; k++) {
            int m = n - k;
            for (auto& a : S[k]) {
                for (auto& b : S[m]) {
                    // Parallel: a + b = (a.p*b.q + b.p*a.q) / (a.q*b.q)
                    long long pn = (long long)a.first * b.second + (long long)b.first * a.second;
                    long long pd = (long long)a.second * b.second;
                    S[n].insert(make_frac(pn, pd));

                    // Series: ab/(a+b) = (a.p*b.p) / (a.p*b.q + b.p*a.q)
                    long long sn = (long long)a.first * b.first;
                    long long sd = (long long)a.first * b.second + (long long)b.first * a.second;
                    S[n].insert(make_frac(sn, sd));
                }
            }
        }
        for (auto& f : S[n]) {
            all_values.insert(f);
        }
    }

    printf("%d\n", (int)all_values.size());
    return 0;
}
